### PR TITLE
Allow picking of a random subfaction.

### PR DESCRIPTION
**Pull request by phrohdoh**
~~Works in ra.~~
~~Crashes in cnc/d2k/ts but I cannot figure out why.~~

~~Help much appreciated!~~

### CODE REVIEW

**Comment by pchote** [comment]
Using magic race names for this feels rather hacky.  It might work out cleaner if you defined the concept of a random class, with each race opting in to one or more of them.

**Comment by WolfGaming** [comment]
We could add a country group trait where you can add all the races you want in it.
For example
```
	CountryGroup:
		Name: Any bla
		Races: bla, bla2, bla3
```

we could even add it to the prereq system so any races in it get it as a provided prereq, but that is optional.

**Comment by Phrohdoh** [comment]
That was actually my initial implement (the name was CountryGroup) but I couldn't get it working.

I'll try again tonight.

**Comment by WolfGaming** [comment]
@Phrohdoh If you have any problems I can help, but I dont have time to code anything myself atm, maybe in a few weeks.

**Comment by LipkeGu** [comment]
```
foreach (var c in selectableCountries)
			{
				var side = c.Side;
				if (side == null)
					continue;

				if (countries.Keys.Contains("random_{0}".F(side.ToLower())))
					continue;

				countries.Add(side, new LobbyCountry { Name = "Any", Side = c.Side });
			}
```

works for me... :)

**Comment by WolfGaming** [comment]
@LipkeGu This however does not support having a race be a part of two groups. Also this would force mods and the like to add sides which makes very little sense if you only have 2 factions.

**Comment by LipkeGu** [comment]
@WolfGaming : its the same logic as in the commit, i added only a check for null to prevent a crashes

**Comment by LipkeGu** [comment]
i changed it again so it more "smarter"
```
foreach (var c in selectableCountries)
{
	var side = c.Side;
	if (side == null)
		continue;

	if (countries.Keys.Contains("random_{0}".F(side.ToLower())) || countries.Keys.Contains(c.Side))
		continue;
				
	countries.Add(side, new LobbyCountry { Name = "Any", Side = c.Side });
}
```

**Comment by Phrohdoh** [comment]
I'm working on this again but would like input on the flag icons created here.
Yay/nay?

**Comment by Unit158** [comment]
You could also try just the try the any flag with the grey bits replaced by the side's respective colour (i.e., a question mark on a blue flag for Allies). Otherwise, it looks fine to me.

**Comment by Phrohdoh** [comment]
Meh, afterall I prefer the current flags personally.

**Update:**
The remaining steps are:
• rename oramap -> zip
• unzip
• `s/"Race: Random"/"Race: random"/`
• rezip
• rename zip -> oramap.

**Comment by Phrohdoh** [comment]
Updated the maps.
Ready for review.

File: OpenRA.Game/Player.cs [hunk 1]
```diff
@@ -46,14 +46,31 @@ public class Player : IScriptBindable, IScriptNotifyBind, ILuaTableBinding, ILua
...
+					return selectableCountries.Random(r);
+
+				var group = world.WorldActor.Info.Traits.WithInterface<CountryGroupInfo>()
+					.FirstOrDefault(g => g.Race == race);
+
+				if (group == null)
+					throw new YamlException("No CountryGroup found with Race `{0}` on map `{1}`.".F(race, world.Map.Title));
+
+				var shuffled = group.RaceMembers.Shuffle(r).ToArray();
+				selected = selectableCountries.First(c => c.Race == shuffled[0]);
```
Comment (RoosterDragon) [inline OpenRA.Game/Player.cs]: Looks like Random would suffice here (unless I'm missing something).
```
var randomRace = group.RaceMembers.Random(r);
selected = selectableCountries.First(c => c.Race == randomRace);
```

**Comment by RoosterDragon** [comment]
This seems to be making the linter fall over (check the Travis logs).

**Comment by pchote** [comment]
@Phrohdoh: sorry to break it to you, but updating the map yaml by hand isn't sufficient, as that doesn't account for the hundreds of maps on the resource site and any custom installed/created maps that players may have (as well as just being really inefficient).  The standard way to do this is to bump the yaml map format and add an on-load check/fix/save.

**Comment by Phrohdoh** [comment]
@RoosterDragon that is due to the `Random` vs `random` yaml changes in maps.

File: OpenRA.Game/Player.cs (hunk 1 above)
Comment (Phrohdoh) [inline OpenRA.Game/Player.cs]: You are right, thank you sir.

**Comment by Phrohdoh** [comment]
Updated (bumped MapFormat suggestion from @pchote).
Fixed a rare OutOfIndex introduced by older versions of this PR's code.

File: OpenRA.Game/Traits/World/CountryGroup.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+﻿// #region Copyright & License Information
+// /*
+//  * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+//  * This file is part of OpenRA, which is free software. It is made
+//  * available to you under the terms of the GNU General Public License
+//  * as published by the Free Software Foundation. For more information,
+//  * see COPYING.
+//  */
+// #endregion
```
Comment (Mailaender) [inline OpenRA.Game/Traits/World/CountryGroup.cs]: The `// ` are a bit strange in front of multi-line comments. I know that MonoDevelop does them when you use a templated header which seems to be a bug.

**Comment by abcdefg30** [comment]
Needs a rebase.

**Comment by Phrohdoh** [comment]
Rebased and went with @Unit158's suggestion about the flags (should I change the outline color too?).
The flags aren't as pretty but they do get the point across better.

Fixed some typos (late night hacking).

File: OpenRA.Game/Traits/World/CountryGroup.cs (hunk 2 above)
Comment (Phrohdoh) [inline OpenRA.Game/Traits/World/CountryGroup.cs]: Fixed. Thanks I always mean to remove this before pushing but obviously forgot here.

**Comment by Phrohdoh** [comment]
@Mailaender Could you please restart the AppVeyor build if possible?

**Comment by Mailaender** [comment]
https://github.com/OpenRA/OpenRA/pull/7408 will make this less dependent on SourceForge mirror availability.

**Comment by penev92** [comment]
We really should start moving away from the Country/Race/CountryRaceFactionHouse names.
"Faction" is generic enough to combine all the others and we should start using that everywhere.
While I'm not saying it should be done in this PR, this is a nice place to start.

File: OpenRA.Game/Network/Session.cs [hunk 3]
```diff
@@ -102,7 +102,7 @@ public static Client Deserialize(MiniYaml data)
 			public int Index;
 			public HSLColor PreferredColor; // Color that the client normally uses from settings.yaml.
 			public HSLColor Color; // Actual color that the client is using. Usually the same as PreferredColor but can be different on maps with locked colors.
-			public string Country;
+			public string CountryRace;
```
Comment (penev92) [inline OpenRA.Game/Network/Session.cs]: `Country` to `Faction` please.
You are renaming this anyway, so might as well do it right the first time.

File: OpenRA.Game/Traits/World/CountryGroup.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+		[Desc("Used for UI flags and other race-specific things.",
+		"Must not be shared with any other Countrys or CountryGroups.")]
+		public readonly string Race = null;
+
+		[Desc("Races that are a member of this group.")]
+		public readonly string[] RaceMembers = { };
+	}
+
+	public class CountryGroup { }
```
Comment (penev92) [inline OpenRA.Game/Traits/World/CountryGroup.cs]: `CountryGroup` -> `FactionGroup` please.

File: OpenRA.Game/Traits/World/CountryGroup.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	{
+		[Desc("String displayed in the lobby dropdown.")]
+		public readonly string DisplayName = null;
+
+		[Desc("Used for UI flags and other race-specific things.",
+		"Must not be shared with any other Countrys or CountryGroups.")]
+		public readonly string Race = null;
+
+		[Desc("Races that are a member of this group.")]
+		public readonly string[] RaceMembers = { };
```
Comment (penev92) [inline OpenRA.Game/Traits/World/CountryGroup.cs]: "FactionMembers", perhaps?

File: OpenRA.Game/Traits/World/CountryGroup.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+{
+	[Desc("Used for random country selection.")]
+	public class CountryGroupInfo : TraitInfo<CountryGroup>
+	{
+		[Desc("String displayed in the lobby dropdown.")]
+		public readonly string DisplayName = null;
+
+		[Desc("Used for UI flags and other race-specific things.",
+		"Must not be shared with any other Countrys or CountryGroups.")]
+		public readonly string Race = null;
```
Comment (penev92) [inline OpenRA.Game/Traits/World/CountryGroup.cs]: No idea about this one, but it shouldn't be `Race`.

File: OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs [hunk 7]
```diff
@@ -142,9 +142,15 @@ void CloseWindow()
 			colorPreview = lobby.Get<ColorPreviewManagerWidget>("COLOR_MANAGER");
 			colorPreview.Color = Game.Settings.Player.Color;
 
-			countries.Add("random", new LobbyCountry { Name = "Any" });
-			foreach (var c in modRules.Actors["world"].Traits.WithInterface<CountryInfo>().Where(c => c.Selectable))
-				countries.Add(c.Race, new LobbyCountry { Name = c.Name, Side = c.Side, Description = c.Description });
+			var worldTraits = modRules.Actors["world"].Traits;
+
+			var countryGroupes = worldTraits.WithInterface<CountryGroupInfo>();
```
Comment (penev92) [inline OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs]: "GroupEs" ?
English, please :D

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs]: derp

File: OpenRA.Game/Traits/World/CountryGroup.cs (hunk 6 above)
Comment (Phrohdoh) [inline OpenRA.Game/Traits/World/CountryGroup.cs]: Sure it should be. It is used in the same way the other `Race`s are used (i.e. flags).

Comment (penev92) [inline OpenRA.Game/Traits/World/CountryGroup.cs]: Nobody said they are correct.

Comment (Phrohdoh) [inline OpenRA.Game/Traits/World/CountryGroup.cs]: Okay, well they are that and I won't change this to match what those aren't.

Comment (pchote) [inline OpenRA.Game/Traits/World/CountryGroup.cs]: @penev92: you may not agree with it, but it is the convention used by OpenRA (and it's used in a lot of places).  This is not the place to try and get it changed.

Comment (pchote) [inline OpenRA.Game/Traits/World/CountryGroup.cs]: Actually... the conventional use of `Race` is actually what `RaceMembers` below is.  I agree that this probably shouldn't be called `Race`.  The description doesn't really clarify the matter, though.

**Comment by penev92** [comment]
I believe that OpenRA has outgrown "Countries" years ago.
We have countries in RA, factions in TD/TS, Houses in D2k. Unofficial mods may include "civilizations" ( @Phrohdoh should know what I'm talking about) and "species".
The only term that is generic enough for all those is "Factions".
The translation to that should start somewhere, and I think naming the new additions you make with this PR in that style is a good start. Otherwise you are just creating more work for whoever decides to do it somewhere in the future.
Sure, it would be even less consistent for a while, with an additional term added to the mess, but the current state of things is inconsistent enough.

On a side note I reviewed it and it works. My only notes are about the naming. But from what I hear on IRC, the `Random` -> `random` renaming might break thirdparty maps?

**Comment by pchote** [comment]
@penev92 the Race and Country terminology is used (consistently) through every part of the engine: from the low level map format and sound code through to the UI rendering and traits.  I disagree completely with the assertion that this is an appropriate place to start changing things.

File: OpenRA.Game/GameInformation.cs [hunk 8]
```diff
@@ -114,7 +114,7 @@ public void AddPlayer(OpenRA.Player runtimePlayer, Session lobbyInfo)
 				Color = runtimePlayer.Color,
 				Team = client.Team,
 				SpawnPoint = runtimePlayer.SpawnPoint,
-				IsRandomFaction = runtimePlayer.Country.Race != client.Country,
+				IsRandomFaction = runtimePlayer.Country.Race != client.CountryRace,
```
Comment (pchote) [inline OpenRA.Game/GameInformation.cs]: What exactly is a CountryRace?

Comment (Phrohdoh) [inline OpenRA.Game/GameInformation.cs]: It was used in the same manner as `Race` but I figured someone would get tripped up by that so I just appended `Race` to the field instead.

`Country.Race != Country` is a bit bogus IMO.

It's the best I could do in a bad situation.

Comment (pchote) [inline OpenRA.Game/GameInformation.cs]: Oh, I see.  This is a historical naming fail: `client.Country` actually specifies a race (and probably wants to be renamed to `Race`).

Comment (Phrohdoh) [inline OpenRA.Game/GameInformation.cs]: That I can do.

**Comment by pchote** [comment]
This feature can be implemented with significantly less code changes:
* Add a `RandomRace` field to `CountryInfo`.
* Add three new `Country` definitions to the world actor:
```
 	Country@random:
 		Name: Any
 		Race: Random
 		Description: A random country.
 	Country@randomallies:
 		Name: Any Allied
 		Race: Random
 		Side: Allies
 		Description: A random Allied country.
 	Country@random:
 		Name: Any Soviet
 		Race: Random
  		Side: Soviet
		Description: A random Soviet country.
```
 * Change `Session` to define `Country = "random"` &rarr; `Race = "Random",`.

You then don't need any new traits or map format changes, which makes this significantly less controversial.

**Comment by Phrohdoh** [comment]
Hmm, I'm really not a fan of that approach to this scenario. I feel it is more error prone on the yaml side.

I also don't like `Random` for the race because all other races are lowercase, so I'd make the map changes either way, personally. Up to you to close this now or wait for me to implement and like the above suggestion.

Or someone else can do it.

**Comment by pchote** [comment]
> I feel it is more error prone on the yaml side.

How so? The world yaml definitions are virtually identical to what's in the PR now, except it reuses the same Country trait instead of adding a new one.

**Comment by Phrohdoh** [comment]
Someone may make a random that contains another random.

I had another reason too but it has escaped me now.

**Comment by Phrohdoh** [comment]
Closing per the conversation on IRC.

**Pull request closed without merging**